File: src/TinRoll.Logic/Manager/UserManager.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using TinRoll.Data.Repository;
using TinRoll.Logic.Manager.Interface;
using TinRoll.Logic.Mapper;
using TinRoll.Shared;

namespace TinRoll.Logic.Manager
{
    public class UserManager : IUserManager
    {
        private IUserRepository _userRepo;

        public UserManager(IUserRepository userRepo)
        {
            _userRepo = userRepo;
        }

        public async Task<UserDto> CreateUserAsync(UserDto user)
        {
            var dbUser = UserMapper.ToDb(user);
            var createduser = _userRepo.CreateUser(dbUser);
            return UserMapper.ToDto(createduser);
        }

        public async Task<UserDto> GetUserAsync(int id)
        {
            var dbUser = _userRepo.GetUser(id);
            return UserMapper.ToDto(dbUser);
        }

        public async Task<IEnumerable<UserDto>> GetUsersAsync()
        {
            var dbUsers = _userRepo.GetUsers();
            var users = dbUsers.Select(u => UserMapper.ToDto(u));
            return users;
        }
    }
}

File: src/TinRoll.Logic/Manager/UserManager.fs
namespace TinRoll.Logic.Manager

open TinRoll.Shared
open TinRoll.Data.Repository
open TinRoll.Logic.Mapper
open System.Threading.Tasks

type IUserManager =
    interface
        abstract member GetUserAsync : int -> UserDto Task
        abstract member GetUsersAsync : unit -> UserDto list Task
        abstract member CreateUserAsync : UserDto -> int Task
    end

type UserManager(userRepo: IUserRepository) =
    interface IUserManager with
        member this.GetUserAsync id =
            let asyncOp = async {
                let dbUser = 
                    userRepo.GetUserAsync id
                    |> Async.RunSynchronously
                return match dbUser with 
                    | Some u -> UserMapper.ToDto u
                    | None -> Unchecked.defaultof<UserDto>
            }
            Async.StartAsTask(asyncOp)
        member this.GetUsersAsync () =
            let asyncOp = async {
                return userRepo.GetUsersAsync ()
                |> Async.RunSynchronously
                |> List.map UserMapper.ToDto
            }
            Async.StartAsTask(asyncOp)
        member this.CreateUserAsync user =
            let dbUser = UserMapper.ToDb user
            let asyncOp = async {
                return userRepo.CreateUserAsync dbUser
                       |> Async.RunSynchronously
            }
            Async.StartAsTask(asyncOp)
            